File: java-src/basic/src/gash/app/ClientApp.java
package gash.app;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

import gash.socket.BasicClient;

public class ClientApp {
    public ClientApp() {
    }

    public static void main(String[] args) {
        var myClient = new BasicClient("app", "127.0.0.1", 2000);
        myClient.connect();

        // Check if the client is connected before proceeding
        if (myClient.isConnected()) {
            myClient.join("tricode");

            try (var br = new BufferedReader(new InputStreamReader(System.in))) {
                while (true) {
                    System.out.print("\nenter ('exit' to quit): ");
					
                    var m = br.readLine();
                    if (m == null || "exit".equalsIgnoreCase(m)) {
                        System.out.println("Exiting...");
                        break;
                    }

                    if (!m.isEmpty()) {
						if(!myClient.isConnected())
						{
							System.out.println("You have been disconnected. Would you like to reconnect? (y/n): ");
							String answer = br.readLine();
							if ("y".equalsIgnoreCase(answer)) {
								myClient = new BasicClient("java-client", "127.0.0.1", 2000);
								myClient.connect();
								if (myClient.isConnected()) {
									System.out.println("Reconnected to the server.");
									myClient.join("tricode");
								} else {
									System.out.println("Failed to reconnect to the server.");
									break;
								}
							} else {
								System.out.println("Exiting due to disconnection...");
								break;
							}
						}
						long startTime = System.nanoTime();
                        myClient.sendMessage(m);
						long endTime = System.nanoTime();
						long latency = (endTime - startTime) / 1_000_000;
						System.out.println("Message send latency: " + latency + " ms");
                    } else {
                        System.out.println("Message is empty, try again.");
                    }
                }
            } catch (IOException e) {
                System.err.println("Error reading from input: " + e.getMessage());
            } finally {
                myClient.stop(); // Ensure the client is stopped and resources are freed
            }
        } else {
            System.out.println("Failed to connect to the server.");
        }
    }
}

File: cpp-src/basic/src/apps/ClientApp.cpp
#include <iostream>
#include <string>
#include "socket/client.hpp" // Adjust this include path as needed
#include <chrono>

int main() {
    basic::BasicClient clt("cpp-client", "127.0.0.1", 2000);
    try {
        clt.connect();

        
        std::string message;
        auto start = std::chrono::high_resolution_clock::now();
        do {
            
            std::cout << "\nenter ('exit' to quit): ";
            std::getline(std::cin, message);

            if (message.empty()) {
                std::cout << "Message is empty, try again." << std::endl;
                continue;
            }

            if (message == "exit") {
                std::cout << "Exiting..." << std::endl;
                break;
            }

            std::cout<< " client status "<< clt.isConnected();

            // Use the 'good' variable or an equivalent getter method if 'isConnected' is not available
            if (!clt.isConnected()) { // This assumes 'good' is public. Adjust if it's private and has a getter.
                std::cout << "You have been disconnected. Would you like to reconnect? (y/n): ";
                std::string answer;
                std::getline(std::cin, answer);
                if (answer == "y") {
                    clt.connect();
                    if (!clt.isConnected()) { // Again, assuming 'good' is public.
                        std::cout << "Failed to reconnect to the server." << std::endl;
                        break;
                    }
                } else {
                    std::cout << "Exiting due to disconnection..." << std::endl;
                    break;
                }
            }

            clt.sendMessage(message);
            auto end = std::chrono::high_resolution_clock::now();
            std::chrono::duration<double, std::milli> latency = end - start;
            std::cout << "Latency: " << latency.count() << " ms" << std::endl;
        } while (true);
    } catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << std::endl;
    }

    // 'clt' is in scope here because it was declared before the 'try' block
    clt.stop(); // Ensure the client is stopped and resources are freed regardless of how the loop exits
    return 0;
}
